### PR TITLE
Feature/effect

### PR DESCRIPTION
**Pull request by tp**
cc @HenriBeck 

The change in here is straightforward, but but the usage in app is getting verbose again:

```ts
const executeBasketCallAndUpdateBasketAsync = withEnv(
  async (
    {dispatch, select, run, fork}: OmsEnv,
    bapiRequest: BapiCall<BasketResponseData>
  ): Promise<BasketResponseData> => {
```

Mostly annoying to me is the inner function declaring and the fact that we either loose the function name or would have to repeat it. So in tests we can't put a nice log when no mock is found for a given function.

### CODE REVIEW

**Comment by HenriBeck** [comment]
We can access the function name from `function.name`. This even works for named arrow functions.
I think this will make it easier to use. The only thing which is a little bit complicated with that are forks I think.

I don't think the syntax is too verbose. It makes executing functions inside a saga much easier.

**Comment by tp** [comment]
This PR still has the issue with type inference, as `run(callApi, requestResponseObject)` breaks the type inference.

Initially I worked around this by creating a different type of effect / effect creator, which would need to be run with the `env` and contain the proper type inside itself.

But the downside would be that the call-site would have to written like this:

```
run(callApi(requestResponseObject))
```

Which in itself isn't bad (I would even prefer it to the normal run), but the issue I have with this is inconsistency between a normal function and one that returns an effect. (At least as long as the same `run` is used for both.)

**Comment by HenriBeck** [comment]
Yeah, this will also make it hard to mock effects in testing.

**Comment by tp** [comment]
True. I haven't been able to come up with a pattern that allows for correct type-inference and being testable as well.

They would all rely on the user using the correct helper functions, such that the initial function and parameters are saved for comparison in tests. (But it's very easy to just depend on a parameter via closure from the outer "generator" function, which then can't be tested for.)

Overall it's just not very nice to use in the application code.

I am wondering whether we should just use `env` explicitly, and then we could use `run(f, env)` and it would be implicit that `f` is a child with full access to the store (one of my big dislikes about redux-saga, that one can't identify how complex a child is and whether it can cause store updates at first glance).

being able to write helpers like this (with deferred env injection) would help with correct type inference, but I see no way to (mostly) guarantee that we collect the function and arguments for testing:
image

**Comment by HenriBeck** [comment]
Check out #23

**Comment by HenriBeck** [comment]
True, the issue about not knowing what this function actually might do with the store is there, but I feel like that's also the same `take` for example.

**Comment by tp** [comment]
work in progress:
image

That way `callBapi123` returns the proper type based on the parameter.

Functions that don't accept an env as first parameter / return an EffectCreator fail to type `never`.

The implementation and overwriting a prop on the function is not the nicest, but at least it would fix the usage.

TODO: in the `id` function return a custom function that captures/exposes the parameters, so we can compare them in testing.

File: src/lib/stateBasedTestHelper.ts [hunk 1]
```diff
@@ -73,11 +77,27 @@ type ValueMockBuilder<T> = {
...
 };
 
-type ValueMock<T> = {
-  type: 'select' | 'call' | 'spawn' | 'fork';
-  func: Function;
-  value: T;
-};
+type ValueMock<T> =
+  | {
+      type: 'call';
```
Comment (HenriBeck) [inline src/lib/stateBasedTestHelper.ts]: I think we don't need a call mock anymore

File: src/lib/environment.ts [hunk 2]
```diff
@@ -48,40 +49,75 @@ export class Environment<StateT, ActionT extends Action> {
...
+   * @param effectOrEffectCreator
+   * @param params
+   */
+  public run<T, P extends any[]>(effect: BoundEffect<Environment<StateT, ActionT>, P, T>): T;
+  public run<T, P extends any[]>(effect: SagaFunc<StateT, ActionT, P, T>, ...params: P): T;
+  public run<P extends any[], T>(
+    effectOrEffectCreator: BoundEffect<Environment<StateT, ActionT>, P, T> | SagaFunc<StateT, ActionT, P, T>,
+    ...args: P
+  ): T {
+    // TODO: Add overload for (env: Env, ...) => T
```
Comment (HenriBeck) [inline src/lib/environment.ts]: Missing check for cancellation

File: src/lib/environment.ts [hunk 3]
```diff
@@ -48,40 +49,75 @@ export class Environment<StateT, ActionT extends Action> {
...
+  /**
+   * Runs the given saga as an attached child.
+   * Cancelling the parent will also cancel the child at the next opportunity.
+   *
+   * @param effectOrEffectCreator
+   * @param params
+   */
+  public run<T, P extends any[]>(effect: BoundEffect<Environment<StateT, ActionT>, P, T>): T;
+  public run<T, P extends any[]>(effect: SagaFunc<StateT, ActionT, P, T>, ...params: P): T;
+  public run<P extends any[], T>(
```
Comment (HenriBeck) [inline src/lib/environment.ts]: So the only difference between `run` and `call` is now the first arg being the environment?

I think we should rename one of them because the difference is not clear from the name.

Comment (tp) [reply src/lib/environment.ts]: Yes, that's the difference. Do you have any naming suggestions for these? I want to raise them with the team today.

The main benefit I still see of not further overloading `call` (or whatever name we'll end up with) is that one can gauge the complexity / impact of the call right from the signature, without having to inspect the type whether it receives `Env` or not.

File: src/lib/environment.ts (hunk 2 above)
Comment (tp) [reply src/lib/environment.ts]: 🙏🏻. added.

I think for all the saga behavior (cancellation, forking) the tests are still severely lacking. Seems to work in practice for our limited use case, but we definitely need test cases for the edge cases.
